### ISSUE TITLE
softmax activation in the last layer

### ISSUE THREAD

**Issue by tms1337** [reinforceio/tensorforce#157]
In my custom network builder (tried layered network builder as well) I have this as a last layer

`x = dense(x, size=len(env.actions), activation='softmax', bias=True, l2_regularization=regularization, scope='dense0')`

But when I print action inside the execute() method of the environment I see that the action parts do not sum to 1. Why could this be? I also tried stacking linear layer and then softmax nonlinearity. I also tried 

`x = tf.nn.softmax(x)` after a linear layer

**Comment by lefnire** [comment]
Are you trying to use the softmax layer to output actions? The observations and actions layers (input and output layers of a typical DNN) are handled automatically by TensorForce depending on the action_space and state_space definition of the environment.

For example, if you have an environment which expects 4 inputs for each observation, and outputs one of 3 discrete actions, then you don't actually put 4 or 3 anywhere in your network builder - you build those into your environment definition. In this case, if we wanted a single hidden layer in our DNN, we'd only use `network_builder([dict(type='dense', size=32)])` (1 layer).  That's to say, I don't think you want that final softmax, if I'm understanding correctly. Is that maybe the issue?

**Comment by tms1337** [comment]
@lefnire I need to output a single action that is a vector of certain length where each entry represents the percentage of the resources that are allocated to an asset. So I do not need ti output the discrete action, but rather a vector that sums up to 1. I hope I am explaining it clear enough if not please ask.

**Comment by lefnire** [comment]
Gotcha, I'll let someone else chime in then. Re input/output, the way these RL setups work is on an environment that takes actions based on an observation. Different lingo than supervised/unsupervised ML: an _observation_ is your input/features (it's a vector of numbers), and an _action_ is your output. Because observations/actions are defined in the environment (whether it's a video game or a car or trading bot), TF defers to the environment to provide those bits. That's to say, the input/output layers are handled behind the scenes.

See [CartPole](https://github.com/openai/gym/wiki/CartPole-v0) for example. When you get past quickstart.py in TF, you're going to be building your own custom OpenAI Gym environment which takes actions and provides observations.

The other bit is that each RL model handles the final layer in the network differently. Some are discrete, some are continuous, some are multimodal distributions, some are softmax, etc etc. Those final layers are in the definition of said model, so TF removes that step from you since it's part of the deal.

**Comment by AlexKuhnle** [comment]
Don't have to add anything to @lefnire's explanation -- as explained, network outputs can be considered 'internal embeddings', which are used to parameterize the actual stochastic action policy (which, in turn, is specified by the environment's action requirements). Hence it's not really possible to directly influence the action calculation, but the same model configuration applies to different environments without changes.

You could achieve the type of action you want by doing the softmax 'externally', essentially as part of the environment, while the actions then are the 'logits' of the distribution, arbitrary continuous values. However, it might be equally reasonable to consider this a proper action type, "distribution" or so, which could be integrated into TensorForce.

If this answers the question, feel free to close this issue. :-)
